src/modules/launcher/Plugins/Microsoft.Plugin.Calculator/CalculateHelper.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) Microsoft Corporation
+// The Microsoft Corporation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.Plugin.Calculator
+{
+    public static class CalculateHelper
+    {
+        private static readonly Regex RegBrackets = new Regex(@"[\(\)\[\]]", RegexOptions.Compiled);
+        private static readonly Regex RegValidExpressChar = new Regex(
+            @"^(" +
+            @"ceil|floor|exp|pi|e|max|min|det|abs|log|ln|sqrt|" +
+            @"sin|cos|tan|arcsin|arccos|arctan|" +
+            @"eigval|eigvec|eig|sum|polar|plot|round|sort|real|zeta|" +
+            @"bin2dec|hex2dec|oct2dec|" +
+            @"==|~=|&&|\|\||" +
+            @"[ei]|[0-9]|[\+\-\*\/\^\., ""]|[\(\)\|\!\[\]]" +
+            @")+$", RegexOptions.Compiled);
+
+        public static bool InputValid(string input)
+        {
+            if (string.IsNullOrWhiteSpace(input))
+            {
+                throw new ArgumentNullException(paramName: nameof(input));
+            }
+
+            if (input.Length <= 2)
+            {
+                return false;
+            }
+
+            if (!RegValidExpressChar.IsMatch(input))
+            {
+                return false;
+            }
+
+            if (!IsBracketComplete(input))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        public static bool IsBracketComplete(string query)
```

Comment:
nit:  We should probably either make this private or have an explicit unit test for this method.  I'm good with either, as the logic is still currently being tested in `InputValid_TestValid_WhenCalled`, but I'd like to make sure there is a dedicated unit test if it ends up being called by other modules in the future.

Reply:
I have rewritten the logic. It now handles complexer behavior, like ([)(]) was valid, but is now invalid. See commit.